B站视频下载会出现错误

Windows10
版本是0.11.0
链接是 https://www.bilibili.com/video/BV1cJ411h74g
错误信息如下
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/iawia002/annie/extractors/bilibili.genParts(0xc000517060, 0x40, 0xc00001c270, 0x2b, 0xc0001dad80, 0x0, 0x0, 0x6a80, 0x0)
        /Users/iawia002/go/src/github.com/iawia002/annie/extractors/bilibili/bilibili.go:79 +0x353
github.com/iawia002/annie/extractors/bilibili.bilibiliDownload(0xc00001c270, 0x2b, 0xc000400000, 0x1a81e, 0x0, 0x4d91829, 0x84bf249, 0xc00023de00, 0xc, 0x1, ...)
        /Users/iawia002/go/src/github.com/iawia002/annie/extractors/bilibili/bilibili.go:385 +0x70b
github.com/iawia002/annie/extractors/bilibili.extractNormalVideo(0xc00001c270, 0x2b, 0xc000400000, 0x1a81e, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1, ...)
        /Users/iawia002/go/src/github.com/iawia002/annie/extractors/bilibili/bilibili.go:238 +0x778
github.com/iawia002/annie/extractors/bilibili.(*extractor).Extract(0x1c2a370, 0xc00001c270, 0x2b, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1, 0x0, ...)
        /Users/iawia002/go/src/github.com/iawia002/annie/extractors/bilibili/bilibili.go:294 +0x25d
github.com/iawia002/annie/extractors.Extract(0xc00001c270, 0x2b, 0x0, 0x0, 0x0, 0x1, 0x0, 0xa, 0x0, 0x0, ...)
        /Users/iawia002/go/src/github.com/iawia002/annie/extractors/extractors.go:112 +0x37a
main.download(0xc000238c80, 0xc00001c270, 0x2b, 0x0, 0x0)
        /Users/iawia002/go/src/github.com/iawia002/annie/main.go:266 +0x3d8
main.main.func2(0xc000238c80, 0x0, 0x0)
        /Users/iawia002/go/src/github.com/iawia002/annie/main.go:235 +0x3fd
github.com/urfave/cli/v2.(*App).RunContext(0xc0000576c0, 0x183e4d0, 0xc000018150, 0xc00002c3e0, 0x2, 0x2, 0x0, 0x0)
        /Users/iawia002/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:322 +0x714
github.com/urfave/cli/v2.(*App).Run(...)
        /Users/iawia002/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
        /Users/iawia002/go/src/github.com/iawia002/annie/main.go:254 +0x136f

--- a/extractors/bilibili/bilibili.go
+++ b/extractors/bilibili/bilibili.go
@@ -73,45 +73,6 @@ func genAPI(aid, cid, quality int, bvid string, bangumi bool, cookie string) (st
 	return api, nil
 }
 
-func genParts(dashData *dashInfo, quality int, referer string) ([]*types.Part, error) {
-	parts := make([]*types.Part, 1)
-	if dashData.Streams.Audio == nil {
-		url := dashData.DURL[0].URL
-		_, ext, err := utils.GetNameAndExt(url)
-		if err != nil {
-			return nil, err
-		}
-		parts[0] = &types.Part{
-			URL:  url,
-			Size: dashData.DURL[0].Size,
-			Ext:  ext,
-		}
-
-	} else {
-
-		checked := false
-		for _, stream := range dashData.Streams.Video {
-			if stream.ID == quality {
-				s, err := request.Size(stream.BaseURL, referer)
-				if err != nil {
-					return nil, err
-				}
-				parts[0] = &types.Part{
-					URL:  stream.BaseURL,
-					Size: s,
-					Ext:  "mp4",
-				}
-				checked = true
-				break
-			}
-		}
-		if !checked {
-			return nil, nil
-		}
-	}
-	return parts, nil
-}
-
 type bilibiliOptions struct {
 	url      string
 	html     string
@@ -359,36 +320,17 @@ func bilibiliDownload(options bilibiliOptions, extractOption types.Options) *typ
 	}
 
 	streams := make(map[string]*types.Stream, len(dashData.Quality))
-	for _, q := range dashData.Quality {
-		// Avoid duplicate streams
-		if _, ok := streams[strconv.Itoa(q)]; ok {
-			continue
-		}
-		api, err := genAPI(options.aid, options.cid, q, options.bvid, options.bangumi, extractOption.Cookie)
+	for _, stream := range dashData.Streams.Video {
+		s, err := request.Size(stream.BaseURL, referer)
 		if err != nil {
 			return types.EmptyData(options.url, err)
 		}
-		jsonString, err := request.Get(api, referer, nil)
-		if err != nil {
-			return types.EmptyData(options.url, err)
-		}
-
-		err = json.Unmarshal([]byte(jsonString), &data)
-		if err != nil {
-			return types.EmptyData(options.url, err)
-		}
-		if data.Data.Description == nil {
-			dashData = data.Result
-		} else {
-			dashData = data.Data
-		}
-		parts, err := genParts(&dashData, q, options.url)
-		if parts == nil {
-			continue
-		}
-		if err != nil {
-			return types.EmptyData(options.url, err)
-		}
+		parts := make([]*types.Part, 0, 2)
+		parts = append(parts, &types.Part{
+			URL:  stream.BaseURL,
+			Size: s,
+			Ext:  getExtFromMimeType(stream.MimeType),
+		})
 		if audioPart != nil {
 			parts = append(parts, audioPart)
 		}
@@ -396,13 +338,14 @@ func bilibiliDownload(options bilibiliOptions, extractOption types.Options) *typ
 		for _, part := range parts {
 			size += part.Size
 		}
-		streams[strconv.Itoa(q)] = &types.Stream{
+		id := fmt.Sprintf("%d-%d", stream.ID, stream.Codecid)
+		streams[id] = &types.Stream{
 			Parts:   parts,
 			Size:    size,
-			Quality: qualityString[q],
+			Quality: fmt.Sprintf("%s %s", qualityString[stream.ID], stream.Codecs),
 		}
 		if audioPart != nil {
-			streams[strconv.Itoa(q)].NeedMux = true
+			streams[id].NeedMux = true
 		}
 	}
 
@@ -431,4 +374,12 @@ func bilibiliDownload(options bilibiliOptions, extractOption types.Options) *typ
 		},
 		URL: options.url,
 	}
+}
+
+func getExtFromMimeType(mimeType string) string {
+	exts := strings.Split(mimeType, "/")
+	if len(exts) == 2 {
+		return exts[1]
+	}
+	return "mp4"
 }

--- a/extractors/bilibili/bilibili_test.go
+++ b/extractors/bilibili/bilibili_test.go
@@ -16,19 +16,17 @@ func TestBilibili(t *testing.T) {
 		{
 			name: "normal test 1",
 			args: test.Args{
-				URL:     "https://www.bilibili.com/video/av20203945/",
-				Title:   "【2018拜年祭单品】相遇day by day",
-				Quality: "高清 1080P",
+				URL:   "https://www.bilibili.com/video/av20203945/",
+				Title: "【2018拜年祭单品】相遇day by day",
 			},
 			playlist: false,
 		},
 		{
 			name: "normal test 2",
 			args: test.Args{
-				URL:     "https://www.bilibili.com/video/av41301960",
-				Title:   "【英雄联盟】2019赛季CG 《觉醒》",
-				Size:    70696896,
-				Quality: "高清 1080P",
+				URL:   "https://www.bilibili.com/video/av41301960",
+				Title: "【英雄联盟】2019赛季CG 《觉醒》",
+				Size:  70696896,
 			},
 			playlist: false,
 		},
@@ -37,24 +35,21 @@ func TestBilibili(t *testing.T) {
 			args: test.Args{
 				URL:   "https://www.bilibili.com/bangumi/play/ep167000",
 				Title: "狐妖小红娘：第70话 苏苏智商上线",
-				// Quality: "高清 1080P",
 			},
 		},
 		{
 			name: "bangumi playlist test",
 			args: test.Args{
-				URL:     "https://www.bilibili.com/bangumi/play/ss5050",
-				Title:   "一人之下：第1话 异人刀兵起，道炁携阴阳",
-				Quality: "高清 720P",
+				URL:   "https://www.bilibili.com/bangumi/play/ss5050",
+				Title: "一人之下：第1话 异人刀兵起，道炁携阴阳",
 			},
 			playlist: true,
 		},
 		{
 			name: "playlist test",
 			args: test.Args{
-				URL:     "https://www.bilibili.com/video/av16907446/",
-				Title:   "\"不要相信歌词，他们为了押韵什么都干得出来\"",
-				Quality: "高清 720P",
+				URL:   "https://www.bilibili.com/video/av16907446/",
+				Title: "\"不要相信歌词，他们为了押韵什么都干得出来\"",
 			},
 			playlist: true,
 		},

--- a/extractors/bilibili/types.go
+++ b/extractors/bilibili/types.go
@@ -46,6 +46,9 @@ type dashStream struct {
 	ID        int    `json:"id"`
 	BaseURL   string `json:"baseUrl"`
 	Bandwidth int    `json:"bandwidth"`
+	MimeType  string `json:"mimeType"`
+	Codecid   int    `json:"codecid"`
+	Codecs    string `json:"codecs"`
 }
 
 type dashStreams struct {
@@ -63,7 +66,6 @@ type dashInfo struct {
 	Description []string    `json:"accept_description"`
 	Quality     []int       `json:"accept_quality"`
 	Streams     dashStreams `json:"dash"`
-	DURL        []dURL      `json:"durl"`
 }
 
 type dash struct {

